src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs

```diff
@@ -282,9 +292,15 @@ void IConnectionControl.End(ProduceEndType endType)
             }
         }
 
-        void IConnectionControl.Stop()
+        void IConnectionControl.SetTimeout(long milliseconds)
+        {
+            // Add KestrelThread.HeartbeatMilliseconds extra milliseconds since this can be called right before the next heartbeat.
+            Interlocked.Exchange(ref _timeoutTimestamp, _lastTimestamp + milliseconds + KestrelThread.HeartbeatMilliseconds);
```

Comment:
Can we have a Debug.Assert or something verifying that the _timeoutTimestamp was long.MaxValue previously?